### ISSUE TITLE
Can one determine if an optional inline command argument has been provided

### ISSUE THREAD

**Issue by ToonTalk** [jmoenig/Snap--Build-Your-Own-Blocks#2163]
I tried this

![image](https://user-images.githubusercontent.com/2138696/43575187-1a85d4b8-963e-11e8-8dfa-e3d7aed1ecd9.png)

but the = predicate reported false.

Generally is the following right?

![image](https://user-images.githubusercontent.com/2138696/43575426-b7061a78-963e-11e8-952c-3ce7d1f34a26.png)

Perhaps related to #1204

Generally I'm unclear about best practice regarding using inline commands versus c-shaped for command arguments

**Comment by bromagosa** [comment]
Comparing

**Comment by bromagosa** [comment]
Sorry, accidentally clicked on "Close issue".

I was about to write that there are different criteria for function comparison. Since what you're comparing are two lambdas, they'll always be different from each other even if their code content looks the same, as their context is necessarily different.

On the other hand, if you assigned a function to a variable and then compared it against itself, you'd get "true":

![image](https://user-images.githubusercontent.com/1016697/43606627-31542452-969c-11e8-97eb-35c22da772ca.png)

This also works for copies:

![image](https://user-images.githubusercontent.com/1016697/43606656-469f5bf6-969c-11e8-82e2-b78967311287.png)

**Comment by KingDavid12** [comment]
@ToonTalk Yes, but it takes some javascript code. Here's a helper block that checks if a ring is empty:
![helper](https://user-images.githubusercontent.com/12404374/43607649-09693602-965c-11e8-862d-446ceef5dc06.png)

**Comment by brianharvey** [comment]
About inline vs. C-shaped:  We use inline in primitives, such as RUN, in which we expect the actual input expression to be a variable rather than a constant script.  We use C-shaped in all other contexts, for which the actual input is likely to be a specific constant script.

In general, there is no way to compare two procedures for EQUAL?-ness.  We follow Scheme in this; the idea is that two procedures might well be functionally equivalent even if their texts are different, e.g., if they have different formal parameters but are otherwise the same, or even if the texs are exactly the same, but they were defined in different environments and use a free variable.  If you think of a procedure as representing a function, two procedures might have altogether different algorithms that produce the same result; my standard example is f(x)=2x+6; g(x)=2(x+3).  So we can only realistically test for EQ?-ness.

On our short-term list is a way to get the text of a lambda expression, so you'll be able to test two texts for EQUAL?-ness, at your own risk with respect to defining environments.

**Comment by ToonTalk** [comment]
Good explanations. Thanks all.

I see that if I use a c-shape for optional command arguments that an unspecified argument is 0. Is this something one can depend upon being true in future versions of Snap!?

**Comment by ToonTalk** [comment]
By the way I see that if I change an argument from inline to c-shape then existing code still works just has redundant rings. But if I change a c-shape argument to inline the command is lost. Would it be wrong to ringify it?

**Comment by ToonTalk** [comment]
There is an apparent downside to c-shape arguments. In the following it seems very awkward to give a nice name to arguments. (test and test2 are identical except test2 using c-shape).

![image](https://user-images.githubusercontent.com/2138696/43635061-ea5c0924-9705-11e8-89e6-d5208f40231a.png)

They work equally well with unnamed arguments but it gets pretty ugly to use a nice name for c-shaped commands. Or is there a way?

**Comment by brianharvey** [comment]
Ringifying is probably the right thing.

0 for an empty script is definitely the wrong thing!  I hereby declare that a bug.

**Comment by cycomachead** [comment]
There's a lot of edge cases around changing types. 

Duplicate of #581 
Relates to #843

> I see that if I use a c-shape for optional command arguments that an unspecified argument is 0. Is this something one can depend upon being true in future versions of Snap!?

This feels like an artifact of JS type coercion. :( At least JS's type rules have a spec. 
`<heavy-snark>`There's no spec for Snap! therefore you can rely on nothing. `</heavy-snark>`

I think inline rings are more like default values, unfortunately. It's worth mentioning that an empty command ring will pass the type test of `is a command ?` so if you wanted to use inline inputs, you'd likely need to check using JSFunction.

This is another use case for introspection!
